### PR TITLE
isValueType and isNonValueType

### PR DESCRIPTION
**Pull request by paulp**
This should supersede #1384.  Review by @xeno-by, @retronym.

### CODE REVIEW

**Comment by paulp** [comment]
PLS REBUILD ALL

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
...
+   *  types which can be used as type arguments (e.g. type constructors.)
+   */
+  def isUseableAsTypeArg(tp: Type) = (
+       isInternalTypeUsedAsTypeArg(tp)  // the subset of internal types which can be type args
+    || isHKTypeRef(tp)                  // not a value type, but ok as a type arg
+    || isValueElseNonValue(tp)          // otherwise only value types
+  )
+
+  private def isHKTypeRef(tp: Type) = tp match {
+    case TypeRef(_, sym, Nil) => tp.isHigherKinded
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:5759]: Can type refs with non-empty args return `true` from `tp.isHigherKinded`?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
...
+  }
+  /** Should be called only with types for which a clear true/false answer
+   *  can be given: true == value type, false == non-value type.  Otherwise,
+   *  an exception is thrown.
+   */
+  private def isValueElseNonValue(tp: Type): Boolean = tp match {
+    case tp if isAlwaysValueType(tp)           => true
+    case tp if isAlwaysNonValueType(tp)        => false
+    case AnnotatedType(_, underlying, _)       => isValueElseNonValue(underlying)
+    case SingleType(_, sym)                    => sym.isValue           // excludes packages and statics
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:5811]: Could you please give an example of a static this check excludes?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 3]
```diff
@@ -5732,6 +5744,100 @@ trait Types extends api.Types { self: SymbolTable =>
...
+    case TypeRef(_, _, _) if tp.isHigherKinded => false                 // excludes type constructors
+    case ThisType(sym)                         => !sym.isPackageClass   // excludes packages
+    case TypeRef(_, sym, _)                    => !sym.isPackageClass   // excludes packages
+    case PolyType(_, _)                        => true                  // poly-methods excluded earlier
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:5815]: What's an example of a `PolyType` which is a value type?

**Comment by gkossakowski** [comment]
Cannot be merged. Also, needs review.

**Comment by xeno-by** [comment]
forked in #1415, rebased against fresh trunk, fixed the failing test

**Comment by xeno-by** [comment]
will try to catch Martin to get a review

**Pull request closed without merging**